ARROW-9937: [Rust] [DataFusion] Improved aggregations

jorgecarleitao opened the pull request:
This PR is a proposal to fix 4 issues in our aggregations:

1. averages are incorrect
2. they only support aggregations that can be reduced using a single value (reason for issue 1.)
3. they do not leverage arrow’s aggregate kernels nor memory layout
4. they only support a single column

The proposal is written here: https://docs.google.com/document/d/1n-GS103ih3QIeQMbf_zyDStjUmryRQd45ypgk884LHU/edit?usp=sharing

Its main features:

1. adds a test of a wrong average and fixes it
1. makes `ScalarValue` a nullable dynamic type, which is closer to how `Array` works (it is nullable)
2. Accumulators now know how to be updated from values (partial) and from other accumulators' state (final)
3. Accumulators can now receive more than one column
4. AggregateExec now knows how to serialize aggregators' state into a ArrayRef's, so that they can be passed throughout the execution
5. Aggregations are now always made in two steps: partial (update from values) and full (update from other's states)
6. Aggregations leverage arrow's kernels as much as possible (all aggregates + take + concatenate)

This PR is built on top of 3 PRs that are under review, and thus is only a draft at this point. 

The benchmarks are between -30% and +15% on my computer (AFAIK no SIMD). Given that the computation now always requires two passes, I was sufficiently happy with them. More can be achieved later.

I am still evaluating the reason for the `aggregate_query_group_by`, but given the functionality that it adds, I considered it sufficiently good for some initial discussions, @andygrove , @nevi-me , @alamb @paddyhoran .

The benchmarks were updated to better reflect real data, and the results are as follows:

```
aggregate_query_no_group_by 15 12                                                                            
                        time:   [478.23 us 479.62 us 480.98 us]
                        change: [-29.686% -27.511% -25.784%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 7 outliers among 100 measurements (7.00%)
  2 (2.00%) high mild
  5 (5.00%) high severe

aggregate_query_group_by 15 12                                                                             
                        time:   [2.8689 ms 2.8794 ms 2.8922 ms]
                        change: [+12.971% +13.710% +14.445%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 5 outliers among 100 measurements (5.00%)
  3 (3.00%) high mild
  2 (2.00%) high severe

aggregate_query_group_by_with_filter 15 12                                                                             
                        time:   [2.1803 ms 2.2062 ms 2.2330 ms]
                        change: [-8.2400% -6.7872% -5.3209%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high mild
```

Sorry for the long PR, but this was a relatively difficult PR to achieve, as it required refactoring of some of our most delicate components. I will try to split it in smaller parts to each the review.

alamb commented:
I plan to review this carefully tomorrow morning US eastern time

On Fri, Sep 11, 2020 at 5:36 PM Jorge Leitao <notifications@github.com>
wrote:

> This PR is a proposal to fix 4 issues in our aggregations:
>
>    1. averages are incorrect
>    2. they only support aggregations that can be reduced using a single
>    value (reason for issue 1.)
>    3. they do not leverage arrow’s aggregate kernels nor memory layout
>    4. they only support a single column
>
> The proposal is written here:
> https://docs.google.com/document/d/1n-GS103ih3QIeQMbf_zyDStjUmryRQd45ypgk884LHU/edit#
>
> Its main features:
>
>    1. adds a test of a wrong average and fixes it
>    2. makes ScalarValue a nullable dynamic type, which is closer to how
>    Array works (it is nullable)
>    3. Accumulators now know how to be updated from values (partial) and
>    from other accumulators' state (final)
>    4. Accumulators can now receive more than one column
>    5. AggregateExec now knows how to serialize aggregators' state into a
>    ArrayRef's, so that they can be passed throughout the execution
>    6. Aggregations are now always made in two steps: partial (update from
>    values) and full (update from other's states)
>    7. MergeExec merges batches in a single batch to reduce batch
>    fragmentation
>    8. Aggregations leverage arrow's kernels as much as possible (all
>    aggregates + take + concatenate)
>
> This PR is built on top of 3 PRs that are under review, and thus is only a
> draft at this point.
>
> The benchmarks are between -30% and +15%. Given that the computation now
> always requires two passes, I was sufficiently happy with them. More can be
> achieved later.
>
> I am still evaluating the reason for the aggregate_query_group_by, but
> given the functionality that it adds, I considered it sufficiently good for
> some initial discussions, @andygrove <https://github.com/andygrove> ,
> @nevi-me <https://github.com/nevi-me> , @alamb <https://github.com/alamb>
> @paddyhoran <https://github.com/paddyhoran> .
>
> The benchmarks were updated to better reflect real data, and the results
> are as follows:
>
> aggregate_query_no_group_by 15 12
>
>                         time:   [478.23 us 479.62 us 480.98 us]
>
>                         change: [-29.686% -27.511% -25.784%] (p = 0.00 < 0.05)
>
>                         Performance has improved.
>
> Found 7 outliers among 100 measurements (7.00%)
>
>   2 (2.00%) high mild
>
>   5 (5.00%) high severe
>
>
>
> aggregate_query_group_by 15 12
>
>                         time:   [2.8689 ms 2.8794 ms 2.8922 ms]
>
>                         change: [+12.971% +13.710% +14.445%] (p = 0.00 < 0.05)
>
>                         Performance has regressed.
>
> Found 5 outliers among 100 measurements (5.00%)
>
>   3 (3.00%) high mild
>
>   2 (2.00%) high severe
>
>
>
> aggregate_query_group_by_with_filter 15 12
>
>                         time:   [2.1803 ms 2.2062 ms 2.2330 ms]
>
>                         change: [-8.2400% -6.7872% -5.3209%] (p = 0.00 < 0.05)
>
>                         Performance has improved.
>
> Found 1 outliers among 100 measurements (1.00%)
>
>   1 (1.00%) high mild
>
>
> Sorry for the long PR, but this was a relatively difficult PR to achieve,
> as it required refactoring of some of our most delicate components. I will
> try to split it in smaller parts to each the review.
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/apache/arrow/pull/8172
> Commit Summary
>
>    - Made aggregates support same signatures as functions.
>    - Added test of average.
>    - Added bench.
>    - Added min/max of StringArray
>    - Improved performance of take.
>    - Refactored aggregate expressions to support complex aggregations.
>
> File Changes
>
>    - *M* rust/arrow/src/compute/kernels/aggregate.rs
>    <https://github.com/apache/arrow/pull/8172/files#diff-d429b07b78d669cf4b46fbb939b2c2b4>
>    (69)
>    - *M* rust/arrow/src/compute/kernels/take.rs
>    <https://github.com/apache/arrow/pull/8172/files#diff-f8e948e2041b5ec28e1ed52afa033c7a>
>    (205)
>    - *M* rust/datafusion/Cargo.toml
>    <https://github.com/apache/arrow/pull/8172/files#diff-3a31d76ae91bd469dc925ad19fae784e>
>    (1)
>    - *M* rust/datafusion/benches/aggregate_query_sql.rs
>    <https://github.com/apache/arrow/pull/8172/files#diff-d651493b657e5fbc784373f751eeb0ac>
>    (134)
>    - *M* rust/datafusion/src/execution/context.rs
>    <https://github.com/apache/arrow/pull/8172/files#diff-8273e76b6910baa123f3a25a967af3b5>
>    (59)
>    - *M* rust/datafusion/src/lib.rs
>    <https://github.com/apache/arrow/pull/8172/files#diff-86191acf7b8e35ce12a84dbb4c4aa80a>
>    (1)
>    - *M* rust/datafusion/src/logical_plan/mod.rs
>    <https://github.com/apache/arrow/pull/8172/files#diff-9922d86c805a8dc858387eb47612caae>
>    (185)
>    - *M* rust/datafusion/src/optimizer/filter_push_down.rs
>    <https://github.com/apache/arrow/pull/8172/files#diff-4de075a6e9df08cd9c9f37333c5dffb1>
>    (14)
>    - *M* rust/datafusion/src/optimizer/projection_push_down.rs
>    <https://github.com/apache/arrow/pull/8172/files#diff-6d2de6d63c825bb8b553d876e2f8a0db>
>    (2)
>    - *M* rust/datafusion/src/optimizer/utils.rs
>    <https://github.com/apache/arrow/pull/8172/files#diff-541a57a1eedc375627a8b9ef9bb82a99>
>    (4)
>    - *A* rust/datafusion/src/physical_plan/aggregates.rs
>    <https://github.com/apache/arrow/pull/8172/files#diff-fd1edcd337781e9ec6538ef3c4591f3d>
>    (156)
>    - *M* rust/datafusion/src/physical_plan/common.rs
>    <https://github.com/apache/arrow/pull/8172/files#diff-a0bf07ce0c251f5bb6acbaa1981c6344>
>    (124)
>    - *M* rust/datafusion/src/physical_plan/expressions.rs
>    <https://github.com/apache/arrow/pull/8172/files#diff-a98d5d588d3c5b525c6840271a5bdddc>
>    (1460)
>    - *M* rust/datafusion/src/physical_plan/filter.rs
>    <https://github.com/apache/arrow/pull/8172/files#diff-5a5d8d45d9149e76b5f98ba90b32ec2b>
>    (7)
>    - *M* rust/datafusion/src/physical_plan/functions.rs
>    <https://github.com/apache/arrow/pull/8172/files#diff-58faffded5f7be9bf6b08dffbcd9eb8b>
>    (18)
>    - *M* rust/datafusion/src/physical_plan/hash_aggregate.rs
>    <https://github.com/apache/arrow/pull/8172/files#diff-a088cd6cbf2e53a43867b69737c485fe>
>    (796)
>    - *M* rust/datafusion/src/physical_plan/mod.rs
>    <https://github.com/apache/arrow/pull/8172/files#diff-444c9d0d823509bd21c36adcf5a58a59>
>    (68)
>    - *M* rust/datafusion/src/physical_plan/planner.rs
>    <https://github.com/apache/arrow/pull/8172/files#diff-8a7abbf6e4953d964b49b836c832b478>
>    (86)
>    - *M* rust/datafusion/src/physical_plan/type_coercion.rs
>    <https://github.com/apache/arrow/pull/8172/files#diff-282dd66764bc913288d68b8c748ca1c0>
>    (18)
>    - *A* rust/datafusion/src/scalar.rs
>    <https://github.com/apache/arrow/pull/8172/files#diff-d2edb06c8da67de0a7f8ae35e7e7e835>
>    (232)
>    - *M* rust/datafusion/src/sql/planner.rs
>    <https://github.com/apache/arrow/pull/8172/files#diff-21b952259e12c1adf2098d9a797caef1>
>    (76)
>    - *M* rust/datafusion/src/test/mod.rs
>    <https://github.com/apache/arrow/pull/8172/files#diff-fb5577d6be40dc90593d523c89fa0f46>
>    (16)
>    - *M* rust/datafusion/src/test/variable.rs
>    <https://github.com/apache/arrow/pull/8172/files#diff-4ab82593dbaf24ea43bad04292f79a1f>
>    (6)
>    - *M* rust/datafusion/src/variable/mod.rs
>    <https://github.com/apache/arrow/pull/8172/files#diff-a0eb1850b19b4e304a0345696bffed77>
>    (2)
>
> Patch Links:
>
>    - https://github.com/apache/arrow/pull/8172.patch
>    - https://github.com/apache/arrow/pull/8172.diff
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/arrow/pull/8172>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADXZMLPBDZORLFYARREPXDSFKJ45ANCNFSM4RIMHT7Q>
> .
>

jorgecarleitao commented on rust/datafusion/benches/aggregate_query_sql.rs:
```diff
@@ -39,72 +46,105 @@ fn aggregate_query(ctx: &mut ExecutionContext, sql: &str) {
+                        .map(
+                            // the 4 here is the number of different keys.
+                            // a higher number increase sparseness
+                            |i| format!("hi{}", i % 4),
```
This should be random, i%4 is quite predictable.

jorgecarleitao commented on rust/datafusion/benches/aggregate_query_sql.rs:
```diff
@@ -39,72 +46,105 @@ fn aggregate_query(ctx: &mut ExecutionContext, sql: &str) {
+    let batch_size = 2048; // 2^11
+    let mut ctx = create_context(partitions_len, array_len, batch_size).unwrap();
+
+    c.bench_function("aggregate_query_no_group_by 15 12", |b| {
```
11, not 12.

jorgecarleitao commented on rust/datafusion/src/physical_plan/common.rs:
```diff
@@ -205,3 +213,91 @@ pub fn get_scalar_value(array: &ArrayRef, row: usize) -> Result<Option<ScalarVal
+            DataType::Boolean => {
+                Ok(Arc::new(BooleanArray::from(vec![] as Vec<bool>)) as ArrayRef)
+            }
+            _ => Err(ExecutionError::NotImplemented(format!(
```
So far this was enough because we do not have aggregations with other types, but this is needed because some batches can have no entries, in which case we need to build an empty record batch.

jorgecarleitao commented on rust/datafusion/src/physical_plan/expressions.rs:
```diff
@@ -97,766 +104,712 @@ pub fn col(name: &str) -> Arc<dyn PhysicalExpr> {
 /// SUM aggregate expression
 #[derive(Debug)]
 pub struct Sum {
+    name: String,
```
`AggregateExpr` has this information with them now because it allows them to `create_accumulator` without access to the input schema. This is helpful because, on the second pass, we need to create accumulators on the fly, and the input_schema of the second pass is different, as it now corresponds to the schema 

`[group1, group2, agg1_state1, agg1_state2, ...]`

jorgecarleitao commented on rust/datafusion/src/physical_plan/expressions.rs:
```diff
@@ -97,766 +104,712 @@ pub fn col(name: &str) -> Arc<dyn PhysicalExpr> {
     }
 }
 
-macro_rules! sum_accumulate {
```
This was operating on a row-by-row basis, which was replaced by a batch operation using `compute::sum`.

jorgecarleitao commented on rust/datafusion/src/physical_plan/expressions.rs:
```diff
@@ -97,766 +104,712 @@ pub fn col(name: &str) -> Arc<dyn PhysicalExpr> {
-        for row in 0..array.len() {
-            self.accumulate_scalar(get_scalar_value(array, row)?)?;
-        }
+    fn merge(&mut self, states: &Vec<ArrayRef>) -> Result<()> {
```
This is the prime example of this PR: the merge here uses two states to change two states from the accumulator.

jorgecarleitao commented on rust/datafusion/src/physical_plan/expressions.rs:
```diff
@@ -1835,88 +1797,18 @@ mod tests {
     }
 
-    #[test]
-    fn sum_contract() -> Result<()> {
```
I removed these because the types are now selected from the signature, and no longer inferred by `AggregateExpr`.

alamb replied:
makes sense

jorgecarleitao commented on rust/datafusion/src/execution/context.rs:
```diff
@@ -848,6 +848,24 @@ mod tests {
     }
 
+    #[test]
+    fn aggregate_grouped_empty() -> Result<()> {
```
A test for grouped with an empty result

jorgecarleitao commented on rust/datafusion/src/execution/context.rs:
```diff
@@ -1147,6 +1165,41 @@ mod tests {
+            .expect("failed to cast version");
+        assert_eq!(values.len(), 1);
+        // avg(1,2,3,4,5) = 3.0
+        assert_eq!(values.value(0), 3.0_f64);
```
This test fails in master, with 3.25 != 3.0

jorgecarleitao commented on rust/datafusion/src/physical_plan/hash_aggregate.rs:
```diff
@@ -276,117 +272,91 @@ impl RecordBatchReader for GroupedHashAggregateIterator {
-                Vec::with_capacity(batch.num_rows());
-
+            // 1.1 construct the key from the group values
+            // 1.2 construct/update the mapping key -> indexes (on the batch) used to `take` values from the batch in a single operation
```
wrong comment, 1.2 and 1.3 are together.

jorgecarleitao commented on rust/datafusion/src/physical_plan/hash_aggregate.rs:
```diff
@@ -709,64 +659,111 @@ fn create_key(
-    use crate::test;
+    use crate::physical_plan::{common, memory::MemoryExec};
+
+    fn some_data() -> ArrowResult<(Arc<Schema>, Vec<RecordBatch>)> {
```
I placed controlled data here so that we cab easily check the result

jorgecarleitao commented on rust/datafusion/src/physical_plan/planner.rs:
```diff
@@ -218,32 +217,18 @@ impl DefaultPhysicalPlanner {
-                )?;
+                )?);
 
-                if initial_aggr.output_partitioning().partition_count() == 1 {
```
This seems major for a single partition.

alamb replied:
I a not sure I understand what you are getting at with this comment

jorgecarleitao replied:
Sorry. What I meant is that compared to the previous implementation, we double the amount of work, even when we did not need because there is a single partition.

IMO we should handle this in another way, e.g. via a physical optimizer that re-writes the partial aggregation by a final aggregation when the input is a single partition.

andygrove replied:
I agree, the planner/optimizer should prevent this.

jorgecarleitao commented on rust/datafusion/src/scalar.rs:
```diff
@@ -0,0 +1,232 @@
+#[derive(Clone, PartialEq)]
+pub enum ScalarValue {
+    /// null value
+    Null,
```
no longer needed?

alamb commented on rust/arrow/src/compute/kernels/aggregate.rs:
```diff
@@ -19,9 +19,42 @@
+        let mut has_value = false;
+        let data = $array.data();
+
+        if null_count == 0 {
```
so the idea here is that the code is faster if we don't have to check each element for nulls?

jorgecarleitao replied:
yes. Generally, operations on non-null fields are faster because there isn't an unpredictable branch on the loop. This check removes that unpredictable branch altogether when there are no nulls.

We could probably still get some more juice by using some of the vertical operations supported by packed_simd.

alamb commented on rust/datafusion/src/physical_plan/aggregates.rs:
```diff
@@ -103,42 +103,54 @@ pub fn create_aggregate_expr(
-        ),
+        AggregateFunction::Min | AggregateFunction::Max => {
+            let mut valid = vec![DataType::Utf8, DataType::LargeUtf8];
+            valid.extend_from_slice(NUMERICS);
```
This is fancy -- do I correctly understand that it allows multiple columns in an aggregate expression (like `MAX(c1, c2)...`)?

jorgecarleitao replied:
I have not though about that, but that is an interesting idea 👍 

In this PR, `max` continues to only support a single column, which we select in [this line](https://github.com/apache/arrow/pull/8172/files#diff-a98d5d588d3c5b525c6840271a5bdddcR571).

This PR does enable us to create aggregate functions with more than one argument, and therefore this allows that option if we wish so. My initial thinking was supporting aggregate functions of more arguments just to support things like `covariance` and `correlation`, but now that you mention, we can do a lot of other things also. Another one is count distinct over N columns.

alamb commented on rust/datafusion/src/physical_plan/expressions.rs:
```diff
@@ -125,192 +128,188 @@ pub fn sum_return_type(arg_type: &DataType) -> Result<DataType> {
+        let array = $NEW_VALUES.as_any().downcast_ref::<$ARRAYTYPE>().unwrap();
+        let delta = compute::sum(array);
+        if $OLD_VALUE.is_none() {
+            ScalarValue::$SCALAR(delta.and_then(|e| Some(e as $TYPE)))
```
👍

alamb reviewed (commented):
I didn't fully grok / review all the code in this PR, but at a high level I like the structure. Nice work @jorgecarleitao 

The idea of two phase aggregates (aka compute average not by averaging averages, but by computing sums of sums and counts of counts and then dividing at the end) 👍 . 

The idea of using arrow compute kernels also seems good to me

I think I saw that this PR adds support for multiple columns in an aggregate expression (like `MAX(c1, c2)...`)? Is that true? 


Thoughts on how to split this up for easier review:

Perhaps https://github.com/apache/arrow/pull/8172/commits/93913d71e0736d41057182f686e8c4ea6210557c (benchmark) could be pulled into its own PR

You maybe could also split out the move of the code for scalar functions to `scalar.rs` in its own PR (with no functional changes) which would be an easy review and reduce the size of this PR substantially.

Then a follow on PR to add support for Nullable scalar value constants

I hear you regarding the PR backup and juggling several in-flight PRs, which is painful. Which ones are you waiting on? Do they need review? Perhaps I can re-review them or @andygrove  can provide some better visibility into what the path is for getting them approved / rejected.

alamb commented:
BTW https://docs.google.com/document/d/1n-GS103ih3QIeQMbf_zyDStjUmryRQd45ypgk884LHU/edit# is a good read too -- thank you for writing that down

andygrove commented:
I did a quick skim through of this and LGTM so far. I will have more time available during the coming week to review more thoroughly and pull the PR locally for some testing.

jorgecarleitao commented:
@alamb , I tried to take the ScalarValue out of the PR for over 2hs, but I was unable: there is too much dependency on it on the current accumulators, and it would be a lot of new code that would then be again removed by this PR.

alamb commented:
Thanks @jorgecarleitao  -- I don't feel strongly about the ScalarValue thign

andygrove commented on rust/datafusion/src/sql/planner.rs:
```diff
@@ -343,7 +343,7 @@ impl<'a, S: SchemaProvider> SqlToRel<'a, S> {
             Some(ref limit_expr) => {
                 let n = match self.sql_to_rex(&limit_expr, &input.schema())? {
-                    Expr::Literal(ScalarValue::Int64(n)) => Ok(n as usize),
+                    Expr::Literal(ScalarValue::Int64(n)) => Ok(n.unwrap() as usize),
```
Could we math on `Some(n)` or add `if n.is_some()` on the match arm to be sure we aren't calling `unwrap` on ` None`?

andygrove commented:
I've been testing with the TPC-H benchmark this morning.

Latest from master:

```
Running benchmarks with the following options: TpchOpt { query: 1, debug: false, iterations: 3, concurrency: 24, batch_size: 4096, path: "/mnt/tpch/parquet/100-240", file_format: "parquet" }
Query 1 iteration 0 took 16167 ms
Query 1 iteration 1 took 16012 ms
Query 1 iteration 2 took 16229 ms
```

This PR:

```
Running benchmarks with the following options: TpchOpt { query: 1, debug: false, iterations: 3, concurrency: 24, batch_size: 4096, path: "/mnt/tpch/parquet/100-240", file_format: "parquet" }
Query 1 iteration 0 took 13638 ms
Query 1 iteration 1 took 13559 ms
Query 1 iteration 2 took 13622 ms
```

That's around 15% speedup :rocket:

andygrove reviewed (approved):
LGTM. Thanks @jorgecarleitao

andygrove commented:
@jorgecarleitao This needs rebasing against master. I ran into compilation issues after merging master into this branch.

```
error[E0308]: mismatched types
   --> datafusion/src/physical_plan/functions.rs:504:33
    |
504 |             ScalarValue::UInt64(1),
    |                                 ^
    |                                 |
    |                                 expected enum `std::option::Option`, found integer
    |                                 help: try using a variant of the expected enum: `Some(1)`
```

jorgecarleitao commented:
@andygrove , that is great news! Really good to know that this stands a stronger benchmark. Thanks a lot for taking the time to run it.

I rebased against master and fixed the unwrap. Let's see if the CI passes.
